### PR TITLE
Fix potential hang issues when encountering illegal external connections

### PR DESCRIPTION
**Pull request by qingyunlei-tencent**
We have addressed the issue raised in #1808 by introducing a new socket state specifically to handle invalid magic values, preventing the program from entering an infinite loop. The PR is based on #1834, and we extend it to cover a key edge case (successful external connection but no magic value or invalid magic field length). Details of the state transitions and handling mechanism are visualized in the diagram below:
image
Test Method as Follows:
Write a shell script that executes nccl-tests alltoall_perf in a loop. During the program runtime, use `naabu` to simulate external connections. When the port used by the proxy thread is connected, it will generate logs and prevent program hangs.
image

Additionally, we have extended the solution to cover an important edge case:
When an external connection is successfully established but either:

1. No magic value is transmitted, or
2. The sent data does not meet the required length for the magic field

A timeout mechanism has been implemented to prevent the program from hanging indefinitely in such cases. This ensures the system can gracefully handle incomplete or malformed connection handshakes, improving overall stability and fault tolerance.
The timeout duration is set to a reasonable default that balances responsiveness with allowance for legitimate network delays, while still preventing permanent hangs in failure scenarios.

### CODE REVIEW

File: src/misc/socket.cc [hunk 1]
```diff
@@ -68,17 +70,48 @@ static ncclResult_t socketProgress(int op, struct ncclSocket* sock, void* ptr, i
       return ncclSuccess;
     } else {
       char line[SOCKET_NAME_MAXLEN+1];
-      WARN("socketProgress: Connection closed by remote peer %s",
-           ncclSocketToString(&sock->addr, line, /*numericHostForm*/0));
+      // Added printing of local listening address and link type information
+      ncclSocketAddress listenAddr;
+      struct sockaddr_in addr;
+      socklen_t len = sizeof(addr);
```
[the remark this answers is not in the document]
Comment (Skylion007) [reply src/misc/socket.cc]:
```suggestion
      constexpr socklen_t len = sizeof(addr);

**Comment by kiskra-nvidia** [comment]
Part of the issue that this PR is targeting has been addressed when we merged #1834.

What remains is the support for a timeout on socket receives. We are considering a narrower approach, that would be limited to the initial socket handshakes, not all receives.
